byron/semantics/executable-spec/src/Control/State/Transition/Generator.hs

```diff
@@ -82,24 +85,72 @@ class STS s => HasTrace s where
     -- ^ Trace length that will be used by 'trace' or 'traceOfLength'.
     -> Gen (Environment s)
 
-  sigGen :: Environment s -> State s -> Gen (Signal s)
+  -- | Generate a signal that __might__ result in the given failure, given an environment and a
+  -- pre-state. Note that the failure generation algorithm can be probabilistic, so the signal is
+  -- not guaranteed to generate the given failure.
+  --
+  -- If 'Nothing' is passed, then the generator should produce a valid signal (according to the
+  -- environment, pre-state, and operational rules of the @sts@ transition system).
+  sigGen
+    :: Maybe (PredicateFailure s)
+    -- ^ Failure that should be generated.
+    -> Environment s
+    -> State s
+    -> Gen (Signal s)
 
   trace
     :: Word64
     -- ^ Maximum length of the generated traces. The actual length will be between 0 and this
     -- maximum.
     -> Gen (Trace s)
-  trace n = traceSigGen (Maximum n) (sigGen @s)
+  trace n = traceWithProfile @s n allValid
+
+  traceWithProfile
+    :: Word64
+    -> TraceProfile s
+    -> Gen (Trace s)
+  traceWithProfile n p = traceSigGenWithProfile (Maximum n) p (sigGen @s)
 
   traceOfLength
     :: Word64
     -- ^ Desired length of the generated trace. If the signal generator can generate invalid signals
     -- then the resulting trace might not have the given length.
     -> Gen (Trace s)
-  traceOfLength n = traceSigGen (Desired n) (sigGen @s)
+  traceOfLength n = traceSigGenWithProfile (Desired n) allValid (sigGen @s)
 
 data TraceLength = Maximum Word64 | Desired Word64
 
+data TraceProfile sts
+  = TraceProfile
+  { proportionOfValidSignals :: Int
+    -- ^ Proportion of valid signals to generate.
+  , proportionOfInvalidSignals :: Int
+    -- ^ Proportion of invalid signals to generate.
+  , failures :: [(Int, PredicateFailure sts)]
+    -- ^ List of failure conditions to try generate when generating an invalid signal, and the
+    -- proportion of each failure.
+  }
```

Comment:
This looks good. I wonder what happens if `[(0, FooBarFailure)]` is passed though. Maybe that is Hedgehog's problem 😆

Reply:
Oh, yeah, good catch. We could add a smart constructor to prevent these kind of values of `hedgehog` chokes on them.